File: leet/0113-path-sum-iii/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if root is None:
            return []
        
        if root.val == sum and root.left is None and root.right is None:
            return [[root.val]]
            
        rs = self.pathSum(root.left, sum - root.val) + self.pathSum(root.right, sum-root.val)
        for ls in rs:
            ls.insert(0, root.val)
            
        return rs

File: leet/0113-path-sum-iii/path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    
    let stack = []
    let ret = []
    
    let dfs = function(node, sum){
        if(node){
            if(node.val == sum && !node.left && !node.right){
                ret.push(stack.concat([node.val]))
                return
            }
            
            stack.push(node.val)
            dfs(node.left, sum-node.val)
            dfs(node.right, sum-node.val)
            stack.pop()
        }
    }
    
    dfs(root, sum)
    return ret
};
